From 47b02953c08a672250cd65377e56585c965207b3 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:37 -0400
Subject: [fs] nfs: teach nfs_lookup_verify_inode to handle LOOKUP_RCU

Message-id: <1411050057-12417-75-git-send-email-steved@redhat.com>
Patchwork-id: 96213
O-Subject: [RHEL7.1 PATCH 74/94] NFS: teach nfs_lookup_verify_inode to handle LOOKUP_RCU
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: NeilBrown <neilb@suse.de>

It fails with -ECHILD rather than make an RPC call.

This allows nfs_lookup_revalidate to call it in RCU-walk mode.

Signed-off-by: NeilBrown <neilb@suse.de>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 1fa1e38447964d6c96b027f9f999ef105d8cf8aa
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/dir.c b/fs/nfs/dir.c
index 7958d41..6b85e78 100644
--- a/fs/nfs/dir.c
+++ b/fs/nfs/dir.c
@@ -1051,6 +1051,8 @@ int nfs_lookup_verify_inode(struct inode *inode, unsigned int flags)
 out:
  return (inode->i_nlink == 0) ? -ENOENT : 0;
 out_force:
+ if (flags & LOOKUP_RCU)
+  return -ECHILD;
  ret = __nfs_revalidate_inode(server, inode);
  if (ret != 0)
   return ret;
@@ -1136,11 +1138,11 @@ static int nfs_lookup_revalidate(struct dentry *dentry, unsigned int flags)
  if (!nfs_is_exclusive_create(dir, flags) &&
      nfs_check_verifier(dir, dentry, flags & LOOKUP_RCU)) {
 
-  if (flags & LOOKUP_RCU)
-   return -ECHILD;
-
-  if (nfs_lookup_verify_inode(inode, flags))
+  if (nfs_lookup_verify_inode(inode, flags)) {
+   if (flags & LOOKUP_RCU)
+    return -ECHILD;
    goto out_zap_parent;
+  }
   goto out_valid;
  }
 
-- 
1.7.1